crates/ruff_python_parser/src/string_token_flags.rs

```diff
@@ -0,0 +1,234 @@
+
+use ruff_text_size::{TextLen, TextSize};
+
+bitflags! {
+    /// [String and Bytes literals]: https://docs.python.org/3/reference/lexical_analysis.html#string-and-bytes-literals
+    /// [PEP 701]: https://peps.python.org/pep-0701/
+    #[derive(Default, Debug, Copy, Clone, PartialEq, Eq, Hash)]
+    struct StringFlagsInner: u8 {
+        /// The string uses double quotes (`"`).
+        /// If this flag is not set, the string uses single quotes (`'`).
+        const DOUBLE = 1 << 0;
+
+        /// The string is triple-quoted:
+        /// it begins and ends with three consecutive quote characters.
+        const TRIPLE_QUOTED = 1 << 1;
+
+        /// The string has a `u` or `U` prefix.
+        /// While this prefix is a no-op at runtime,
+        /// strings with this prefix can have no other prefixes set.
+        const U_PREFIX = 1 << 2;
+
+        /// The string has a `b` or `B` prefix.
+        /// This means that the string is a sequence of `int`s at runtime,
+        /// rather than a sequence of `str`s.
+        /// Strings with this flag can also be raw strings,
+        /// but can have no other prefixes.
+        const B_PREFIX = 1 << 3;
+
+        /// The string has a `f` or `F` prefix, meaning it is an f-string.
+        /// F-strings can also be raw strings,
+        /// but can have no other prefixes.
+        const F_PREFIX = 1 << 4;
+
+        /// The string has an `r` or `R` prefix, meaning it is a raw string.
+        /// F-strings and byte-strings can be raw,
+        /// as can strings with no other prefixes.
+        /// U-strings cannot be raw.
+        const R_PREFIX = 1 << 5;
+    }
+}
+
+impl TryFrom<char> for StringFlagsInner {
+    type Error = String;
+
+    fn try_from(value: char) -> Result<Self, String> {
+        let result = match value {
+            '\'' => Self::empty(),
+            '"' => Self::DOUBLE,
+            'r' | 'R' => Self::R_PREFIX,
+            'u' | 'U' => Self::U_PREFIX,
+            'b' | 'B' => Self::B_PREFIX,
+            'f' | 'F' => Self::F_PREFIX,
+            _ => return Err(format!("Unexpected prefix or quote '{value}'")),
+        };
+        Ok(result)
+    }
+}
+
+const DISALLOWED_U_STRING_PREFIXES: StringFlagsInner = StringFlagsInner::B_PREFIX
+    .union(StringFlagsInner::F_PREFIX)
+    .union(StringFlagsInner::R_PREFIX);
+
+const DISALLOWED_B_STRING_PREFIXES: StringFlagsInner =
+    StringFlagsInner::U_PREFIX.union(StringFlagsInner::F_PREFIX);
+
+const DISALLOWED_F_STRING_PREFIXES: StringFlagsInner =
+    StringFlagsInner::U_PREFIX.union(StringFlagsInner::B_PREFIX);
+
+const DISALLOWED_R_STRING_PREFIXES: StringFlagsInner = StringFlagsInner::U_PREFIX;
+
+#[derive(Default, Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct StringFlags(StringFlagsInner);
+
+impl TryFrom<char> for StringFlags {
+    type Error = String;
+
+    fn try_from(value: char) -> Result<Self, String> {
+        Ok(Self(StringFlagsInner::try_from(value)?))
+    }
+}
+
+impl StringFlags {
```

Comment:
I like that you abstracted away the fact that we use bitflags internall. I wonder if we should take this one step further by:

* Introducing a new `StringPrefix` enum that's `FString`, `RawBytes`, `Raw`, `UString` (similar to the existing `StringKind` enum)
* Reducing the methods on this type. E.g. instead of `string_flags.quote_char()` to `string_flags.quote_style().as_char()` or implement `Display` for `QuoteStyle` and `StringPrefix` (and `StringFlags`?) so that `format!("{prefix}{quotes}", string_flags.prefix(), string_flags.quote_style())` works.

The advantage of exposing the `StringPrefix` is that we can keep the guarantee that you can't construct incorrect invariants without having to use `try` methods but use the more efficient bitflag storage internally. 

That means, the mutation methods would change to `StringFlags::with_prefix(self, prefix: StringPrefix)` and they're guaranteed to never fail. 

Doing this could also help with the lexer because we could then keep the lexer largely unchanged because it would call `StringPrefix::try_from` (or `parse`?) as it used to before and can construct the `StringFlags` by calling `StringFlags::new(prefix, quotes)`